crates/valence_nbt/src/snbt.rs

```diff
@@ -0,0 +1,685 @@
+                list.into_iter().map(|v| v.into_list().unwrap()).collect(),
+            )),
+            Tag::Compound => Ok(List::Compound(
+                list.into_iter()
+                    .map(|v| v.into_compound().unwrap())
+                    .collect(),
+            )),
+            Tag::IntArray => Ok(List::IntArray(
+                list.into_iter()
+                    .map(|v| v.into_int_array().unwrap())
+                    .collect(),
+            )),
+            Tag::LongArray => Ok(List::LongArray(
+                list.into_iter()
+                    .map(|v| v.into_long_array().unwrap())
+                    .collect(),
+            )),
+        }
+    }
+
+    fn parse_list_like(&mut self) -> Result<Value> {
+        self.next();
+        let type_char = self.peek()?;
+        let etype = match type_char {
+            'B' => Tag::Byte,
+            'I' => Tag::Int,
+            'L' => Tag::Long,
+            _ => return self.continue_parse_list().map(|l| l.into()),
+        };
+        self.next();
+        if self.peek()? != ';' {
+            self.push_back(type_char);
+            return self.continue_parse_list().map(|l| l.into());
+        }
+        self.next();
+        self.skip_whitespace();
+        let mut values = vec![];
+        while self.peek()? != ']' {
+            let value = self.parse_element()?;
+            if value.get_type() != etype {
+                return Err(self.make_error(SNBTErrorType::WrongTypeInArray));
+            }
+            values.push(value);
+            self.skip_whitespace();
+            if self.peek()? == ',' {
+                self.next();
+                self.skip_whitespace();
+            } else if self.peek()? != ']' {
+                return Err(self.make_error(SNBTErrorType::ExpectComma));
+            }
+        }
+        self.next();
+        match etype {
+            Tag::Byte => Ok(Value::ByteArray(
+                values.into_iter().map(|v| v.into_byte().unwrap()).collect(),
+            )),
+            Tag::Int => Ok(Value::IntArray(
+                values.into_iter().map(|v| v.into_int().unwrap()).collect(),
+            )),
+            Tag::Long => Ok(Value::LongArray(
+                values.into_iter().map(|v| v.into_long().unwrap()).collect(),
+            )),
+            _ => unreachable!(),
+        }
+    }
+
+    fn parse_primitive(&mut self) -> Result<Value> {
+        macro_rules! try_ret {
+            // Try possible solution until one works
+            ($v:expr) => {{
+                match $v {
+                    Ok(v) => return Ok(v.into()),
+                    Err(_) => (),
+                }
+            }};
+        }
+        let target = self.read_unquoted_string()?;
+        match target
+            .bytes()
+            .last()
+            .ok_or_else(|| self.make_error(SNBTErrorType::ExpectValue))?
+        {
+            b'b' | b'B' => try_ret!(target[..target.len() - 1].parse::<i8>()),
```

Comment:
You can replace `target[..target.len() - 1]` with `target`.

Reply:
We need to pop the last char before `match` and add it back after that to replace `target[..target.len() - 1]` with `target`, which I think the solution is not so clear.

Reply:
Ah, you're right. I misread.